src/error_maps.jl

```diff
@@ -0,0 +1,279 @@
+
+"""
+    ErrorMap
+
+A nonlinear mapping between the space of unit quaternions and three-dimensional
+rotation errors.
+
+These mappings are extremely useful for converting from globally nonsingular 3D rotation
+representations such as `UnitQuaternion` or `RotMatrix3` to a three-parameter error that can
+be efficiently used in gradient-based optimization methods that optimize deviations about
+a current iterate using first or second-order information.
+
+# Usage
+    errmap(v::AbstractVector)  # "forward" map from a 3D error to a `UnitQuaternion`
+    errmap(R::Rotation)        # "inverse" map from a rotation (via `UnitQuaternion`) to a 3D error
+
+where `errmap <: ErrorMap`
+
+# Implemented Maps
+- `CayleyMap`:  Uses `RodriguesParam` as the error representation (default).
+    Goes singular at 180° and does not have a sign ambiguity.
+- `ExponentialMap`: Uses the canonical exponential map from Lie Group theory. Computationally
+    expensive to compute. Exhibits kinematic singularities.
+- `MRPMap`: Uses a scaled `MRP` as the error representation. Singular at 360° but has
+    a sign ambiguity (with the "shadow" MRP set).
+- `QuatVecMap`: Uses the vector part of the quaternions. Cheapest map to compute, but
+    goes singular at 180° and suffers from sign ambiguity.
+- `IdentityMap`: Maps values through directly. Only works with three-parameter rotation
+    representations with the following methods: `R(::SVector{3})` and `SVector(::R)::SVector{3}`
+"""
+abstract type ErrorMap end
+struct CayleyMap <: ErrorMap end
+struct ExponentialMap <: ErrorMap end
+struct MRPMap <: ErrorMap end
+struct QuatVecMap <: ErrorMap end
+struct IdentityMap <: ErrorMap end
+
+const DEFAULT_EMAP = CayleyMap()
+
+# Scalings
+@inline scaling(::Type{ExponentialMap}) = 0.5
+@inline scaling(::Type{QuatVecMap}) = 1.0
+@inline scaling(::Type{CayleyMap}) = 1.0
+@inline scaling(::Type{MRPMap}) = 2.0
+scaling(m::M) where M <: ErrorMap = scaling(M)
+
+# Quaternion Maps
+forward_map(::ExponentialMap, ϕ::AbstractVector) = expm(ϕ/scaling(ExponentialMap))
```

Comment:
These `forward_map` and `inverse_map` have me scratching my head a bit.  It reminds me of the early design we had in `CoordinateTransformations.jl` (or maybe `Proj4.jl`) and I think we ended up scrapping that (memory is hazy though).

Anyway, I think it could be helpful to be more literal here and make `ExponentialMap` objects callable:

```julia
(::ExponentialMap)(ϕ::AbstractVector) = expm(ϕ/scaling(ExponentialMap))

# To handle the inverse:
struct InvExponentialMap
end

inv(::ExponentialMap) = InvExponentialMap()

(::InvExponentialMap)(q::UnitQuaternion) = scaling(ExponentialMap)*logm(q)
```

This has the benefit that things like broadcasting become very natural when you have a map object `f` because applying the mapping is simply a function call. And to apply the inverse, `inv(f)(x)` should have zero overhead.

Reply:
Good suggestion. This is fairly similar to how I had it originally, except that I wasn't using concrete singletons and I relied to multiple dispatch to pick between the forward and inverse maps. Given that rotations devolve to an `SMatrix` I agree that it's safer to be explicit about requesting the inverse map (rotation -> error) vs the forward map (error -> rotation).